PROJ-72 - Network format (#36)
SW805F20/report: sections/sprint-1/networking.tex

IPv6 addresses are usually written as eight 16-bit hexadecimal numbers \cite{socketnetworking}.
The higher-order bits of the 128-address imply the type.
For this project, this should not have a significant impact, and either can be used.

\subsection{Format of data}

> As it is needed to send data over the network for our game, it is required to define the format it is sent as.

Reviewer: ```suggestion
As it is necessary to send data over the network, the format of the data must be decided upon.
```

The network format for the field anchor, player position and goal position is \texttt{0xyyyyxxxxiisstt}, where \texttt{t} is the type, \texttt{s} is the timestamp, \texttt{i} is the ID, \texttt{xxxx} is the \texttt{x} position and \texttt{yyyy} is the \texttt{y} position.
These formats can be seen on \autoref{Tab:anchor-format}, \autoref{Tab:player-format} and \autoref{Tab:player-format}.
These are almost identical, as the same information is required.

> The only difference between them is the type, and the context of the IDs.

Reviewer: ```suggestion
The data sent is almost identical, as the same information is required, the only difference between them is the type and the context of the IDs.
```

> The type is at the end of the hexadecimal, so that it is easy to bitshift, and then switch on which type it is.

Reviewer: ```suggestion
The type is at the end of the hexadecimal so that it is easy to bitshift, and then switch on which type it is.
```

er ikke stor fan af den her sætning, det virker som et ret halvt argument

After the type the timestamp is placed, which is used to detect outdated data.
\\\\

> It was decided to use 2 and 4 hexadecimal, so that it is in byte format and hereby straightforward to work with in \texttt{C\#}.

Reviewer: ```suggestion
It was decided to use a size of respectively 2 and 4 hexadecimal so that it is in byte format and i straightforward to work with in the \texttt{C\#} language.
```

er det her det fulde argument? @jenspetur94

Reviewer 2: Det fulde argument er at 2 og 4 hexadecimaler svarer til 1 og 2 bytes. Det er nemmere at arbejde med i C# fordi at enkelte dele af package har samme størrelse som nogle typer i C#. Derfor kan man bitshifte sådan at det argument i pakken man skal aflæse forrest og så type caste packagen til den type der svarer til størrelsen på argumentet.

Reviewer 2: ![image](https://user-images.githubusercontent.com/22811843/77744661-35a9c300-701a-11ea-8d20-86c2dad46b34.png)

> Given the highest number that can be in two bytes are $2^{16} - 1$, the maximum number for the \texttt{x} and \texttt{y} positions are 65.535. 

Reviewer: ```suggestion
Given the highest number that can be in two bytes are $2^{16} - 1$, the maximum number for the \texttt{x} and \texttt{y} positions are 65.535, which corresponds to a field size of approximately 655 by 655 meters.
```

\begin{table}[h!]
    \centering
    \begin{tabular}{|l|l|l|l|l|}
    \hline
    \begin{tabular}[c]{@{}l@{}}POSY (y)\\ (0-65.535)\end{tabular} & \begin{tabular}[c]{@{}l@{}}POSX (x)\\ (0-65.535)\end{tabular} & \begin{tabular}[c]{@{}l@{}}ID (i)\\ (0-3)\end{tabular} & \begin{tabular}[c]{@{}l@{}}TIMESTAMP (s)\\ (0-255)\end{tabular} & \begin{tabular}[c]{@{}l@{}}TYPE (t)\\ (0)\end{tabular} \\ \hline

>     4 hexadecimal                                                 & 4 hexadecimal                                                 & 2 hexadecimal                                          & 2 hexadecimal                                                 & 2 hexadecimal                                        \\ \hline

Reviewer: Bør skrives som bits, da det er standarden for det

![image](https://user-images.githubusercontent.com/21969475/77646190-57e10980-6f64-11ea-9e00-16089684cf5f.png)

Reviewer 2: bare som lidt ekstra info så svarer 1 hexadecimal til 4 bits.

IPv6 addresses are usually written as eight 16-bit hexadecimal numbers \cite{socketnetworking}.
The higher-order bits of the 128-address imply the type.
For this project, this should not have a significant impact, and either can be used.

\subsection{Format of data}\label{subsec:data-format}
As it is necessary to send data over the network, the format of the data must be decided upon.
The network format for the field anchor, player position and goal position is \texttt{0xyyyyxxxxiisstt}, where \texttt{t} is the type, \texttt{s} is the timestamp, \texttt{i} is the ID, \texttt{xxxx} is the \texttt{x} position and \texttt{yyyy} is the \texttt{y} position.
These formats can be seen on \autoref{Tab:anchor-format}, \autoref{Tab:player-format} and \autoref{Tab:player-format}.
The data sent is almost identical, as the same information is required, the only difference between them is the type and the context of the IDs.
The type is at the right hand side of the hexadecimal so that it is faster to decode and to switch on which type it is.
After the type the timestamp is placed, which is used to detect outdated data.
\\\\

> It was decided to use a size of respectively 16 and 8 bits so that it is straightforward to work with in the \texttt{C\#} language, because it is the same size as some variables in \texttt{C\#} and hereby easy to type cast.

Reviewer 2: ```suggestion
It was decided to use a size of respectively 16 and 8 bits so that it is straightforward to work with in the \texttt{C\#} language, because it is the same size as some variable types in \texttt{C\#} and hereby easy to type cast.
```

> Given the highest number that can be in two bytes are $2^{16} - 1$, the maximum value for the \texttt{x} and \texttt{y} positions are 65.535.

Reviewer 2: ```suggestion
Given the highest number that can be in two bytes is $2^{16} - 1$, the maximum value for the \texttt{x} and \texttt{y} positions is 65.535.
```

These formats can be seen on \autoref{Tab:anchor-format}, \autoref{Tab:player-format} and \autoref{Tab:goal-position-format}.
The data sent is almost identical, as the same information is required, the only difference between them is the type and the context of the IDs.
The \texttt{x} and \texttt{y} position is the location of the player or anchor.
For \autoref{Tab:goal-position-format} the \texttt{x} and \texttt{y} position is the center of the goal, where the size of the goal is set to have a length equal to 20\% of the shortest side of the playing field.
The IDs for \autoref{Tab:anchor-format} are for each anchor.
For \autoref{Tab:player-format} the ID 0 represents the ball and 1-4 represents a player and for \autoref{Tab:goal-position-format} the ID identifies the team.
The type is at the right hand side of the hexadecimal so that it is faster to decode and to switch on which type it is.
After the type the timestamp is placed, which is used to detect outdated data.

> The timestamp is an integer value ranging between 0 and 255 such that when it hits 255 it resets to 0 to be able to continuously send timestamps, regardless of game length.```

Reviewer: ```suggestion
The timestamp is an integer value ranging between 0 and 255 such that when it hits 255 it resets to 0 to be able to continuously send timestamps, regardless of game length.
```

\subsection{Format of data}\label{subsec:data-format}
As it is necessary to send data over the network, the format of the data must be decided upon.
The network format for the field anchor, player position and goal position is \texttt{0xyyyyxxxxiisstt}, where \texttt{t} is the type, \texttt{s} is the timestamp, \texttt{i} is the ID, \texttt{xxxx} is the \texttt{x} position and \texttt{yyyy} is the \texttt{y} position.
These formats can be seen on \autoref{Tab:anchor-format}, \autoref{Tab:player-format} and \autoref{Tab:goal-position-format}.
The data sent is almost identical, as the same information is required, the only difference between them is the type and the context of the IDs.
The \texttt{x} and \texttt{y} position is the location of the player or anchor.

> For \autoref{Tab:goal-position-format} the \texttt{x} and \texttt{y} position is the center of the goal, where the size of the goal is set in unity.

Reviewer: ```suggestion
For \autoref{Tab:goal-position-format} the \texttt{x} and \texttt{y} position is the center of the goal, where the size of the goal is set to have a length equal to 20\% of the shortest side of the playing field.
```

> The IDs for \autoref{Tab:anchor-format} is for each anchor.

Reviewer: ```suggestion
The IDs for \autoref{Tab:anchor-format} are for each anchor.
```

> For \autoref{Tab:player-format} the ID 0 is the ball, where 1-4 is a player and for \autoref{Tab:goal-position-format} the ID identifies what team it is..

Reviewer: ```suggestion
In \autoref{Tab:player-format}, the ID 0 is reserved for the ball and 1 and up indicates a player and for \autoref{Tab:goal-position-format} the ID indicates which team the goal belongs to.
```

The type is at the right hand side of the hexadecimal so that it is faster to decode and to switch on which type it is.
After the type the timestamp is placed, which is used to detect outdated data.

> The timestamp is an integer that represents time from 0-255 and when it hits 255 it loops to 0 again as the newest value.

Reviewer: ```suggestion
The timestamp is an integer value ranging between 0 and 255 such that when it hits 255 it resets to 0 to be able to continuously send timestamps, regardless of game length.```

When implementing this, it is required to consider if some packages are not received.

> A solution to this could be that 0 to 20 is considered as newer timestamps than 235 to 255 as it is very unlikely that 20 packages are not receive 20 times in a row.

Reviewer: ```suggestion
The downside to this approach is since UDP does not guarantee delivery of messages, it may not go straight from 255 to 0 every time.
To counter this, 0 to 20 is considered as newer timestamps than 235 to 255 so that the system is allowed to miss up to 20 messages near the end, to ensure that it does not end up being locked until the values reach the high range again.
```

> This is however not a guarantee.

Reviewer: ```suggestion
```
